### ISSUE TITLE
Add a `refit` parameter to the StackingClassifier and StackingCVClassifier (similar to the EnsembleVoteClassifier)

### ISSUE THREAD

**Issue by konradsemsch** [rasbt/mlxtend#627]
Hi! First of all thanks for an awesome package @rasbt :) Secondly, I wasn't able to find a similar question but I apologize in case it's a copy. It's essentially quite simple:

1) I have some data on which I have divided in into train/ test as well as a repeated CV scheme with a given random seed to have reproducible results

2) I have trained and optimised hyperparameters of both underlying base models through CV

3) I would like now to train a meta learner on top that would be Elastic-Net and I would like to tune it's hyperparameters using CV as well

If I understand correctly, currently I would need to do something like this:

```
stack_meta = LogisticRegression()

stack_pipeline = StackingClassifier(
    classifiers = [grid_search_rf, grid_search_en], 
    meta_classifier = stack_meta, 
    use_probas = True
    )

grid_param_stack = {'meta_classifier__C': [2**i for i in range(-2,1)]}

grid_search_stack = GridSearchCV(estimator = stack_pipeline, 
                                 param_grid = grid_param_stack, 
                                 scoring = ['roc_auc'],
                                 cv = cv_splits,
                                 refit = 'roc_auc',
                                 n_jobs = 4)
```

My expectation was that since I'm providing tuned underlying classifiers, the ```StackingClassifier``` function would simply aggregate hold-out predictions from CV of my base models, and use them as inputs for tuning the final meta learner while using the same CV scheme. That should normally take a matter of a couple of seconds, but it seems that StackingClassifier is reffing the underlying models as well.

How can I achieve what I want with the current implementation of the function? I couldn't find a definitive answer in the documentation. Thanks!

**Comment by rasbt** [comment]
Hi there,

the main reason why this refitting by default is that it is a default in scikit-learn.  If it wouldn't, then `GridSearchCV` wouldn't work (because here, everything has to be refit to the training folds in k-fold cross-validation).

Currently, only the EnsembleVoteClassifier has a `refit` parameter, which would suppress refitting when `refit=False`. The `StackingClassifier` and `StackingCVClassifier` don't have that yet, but it would be good to add that in future.

**Comment by konradsemsch** [comment]
Thanks for a prompt response! Yeah, this feature would be definitely extremely helpful in both cases :)

**Comment by bmreiniger** [comment]
@konradsemsch I think in your example you should use the grid search only inside the meta_classifier.  The base models will still be fit with cross-val predictions (and then refit for future use), but then those predictions will be static across the grid search.

However, being able to "freeze" the base models may still be desirable.  This is already done e.g. in [baikal](https://github.com/alegonz/baikal).

**Comment by konradsemsch** [comment]
What do you mean by 'inside the meta_classifier'? Isn't this what I'm trying to do? Take two models and use a ridge regression meta classier which I'm trying to optimise on top of the outputs of the two underlying models?

**Comment by bmreiniger** [comment]
@konradsemsch , Sorry for being unclear.  I mean that, in your code above, for each value of `C` you'll refit all of the base learners on each fold split.  If you're happy with the search spaces for all the models being independent, rather than searching the cartesian product of all the individual ones, you can just pass the grid search object in as the meta_classifier: 

```
stack_meta = LogisticRegression()
grid_param_meta = {'C': [2**i for i in range(-2,1)]}

grid_search_meta = GridSearchCV(estimator = stack_meta, 
                                param_grid = grid_param_stack, 
                                scoring = ['roc_auc'],
                                cv = cv_splits,
                                refit = 'roc_auc',
                                n_jobs = 4)

stack_pipeline = StackingClassifier(
    classifiers = [grid_search_rf, grid_search_en], 
    meta_classifier = grid_search_meta, 
    use_probas = True
    )

```

AIUI, this will fit the base models just once (per fold split), with their own grid searches if desired.

**Comment by bmreiniger** [comment]
As for implementation, see also the keras-style freezing of layers with parameter "trainable" ([docs](https://keras.io/getting-started/faq/#how-can-i-freeze-keras-layers)) or sklearn's "prefit" parameter (e.g. in [SelectFromModel](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFromModel.html) and [CalibratedClassifierCV](https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html)).

**Comment by konradsemsch** [comment]
Ok, I get your point right now - that makes much more sense and is pretty much what I've been searching for! So actually the 'refit' option wouldn't be really needed here, it's just a matter of how to build up the pipeline :) Thanks for clarifying this!

@rasbt - perhaps the examples on the wiki page should be slightly amended to account for this?

**Comment by nilichen** [comment]
I'd like to take a look at this.

**Comment by rasbt** [comment]
Awesome, thanks!

**Comment by nilichen** [comment]
There's a contradictory parameter in StackingClassifier and StackingCVClassifier: `use_clones`. Basically we cannot have `refit=False` with `use_clones=True` (which is the default). I would assume not a lot of users are aware of `use_clones`, so should we replace it with `refit`?  Therefore by default, it always uses clone, just like EnsembleVoteClassifier? 
```
    use_clones : bool (default: True)
        Clones the classifiers for stacking classification if True (default)
        or else uses the original ones, which will be refitted on the dataset
        upon calling the `fit` method. Hence, if use_clones=True, the original
        input classifiers will remain unmodified upon using the
        StackingClassifier's `fit` method.
        Setting `use_clones=False` is
        recommended if you are working with estimators that are supporting
        the scikit-learn fit/predict API interface but are not compatible
        to scikit-learn's `clone` function.
```

**Comment by rasbt** [comment]
Thanks for bringing this up. It's been a long time since I thought about this ... 

Like you said, it gets complicated if `use_clones=True` and `refit=False`. There are some possibilities for handling this. For example,

- If `refit=False`, use original classifier w/o refitting and ignore option for `use_clones`
- If `refit=True`,  choose either original classifiers for refitting or clone the original classifiers before refitting (depending on the option chosen for `use_clones`).

I think that this gets complicated from user perspective though. But this would be the only way to account for all use cases.

Maybe calling it `fit_base_estimators` would be also good for more clarity (to make clear that the meta-classifier is not affected).

So, the default option would be 

- `fit_base_estimators=True`
- `use_clones=True`

What do you think?

**Comment by nilichen** [comment]
That's reasonable. On a second thought, I don't think it's possible to have `fit_base_estimators=False` for  StackingCVClassifier and  StackingCVRegressor as they must call `cross_val_predict` during fit. 

As such, I will implement `fit_base_estimators` and `use_clones` for EnsembleVoteClassifier, StackingClassifier and StackingRegressor so that they have consistent interface.

**Comment by rasbt** [comment]
Good point! So far the CV variants, StackingCVClassifier and StackingCVRegressor, we wouldn't have a use_clones parameter. I just see this is currently implemented though. Not sure what happens if it is set to `use_clones=False` -- it will probably fail like you said (weird that we don't have a unit test for that then)

**Comment by nilichen** [comment]
I did try `use_clones=False`, actually it won't fail as it will still fit the data, with the side effect of modifying the original input classifier. My point would be if having `fit_base_estimators=False` in StackingCVClassifer, it will not have effect during fitting the `meta_classifier`, https://github.com/rasbt/mlxtend/blob/7ed21f6c5df108102c26d8cdb3c4b899e007be61/mlxtend/classifier/stacking_cv_classification.py#L238 but we can still save half of the computation time by skipping this part: https://github.com/rasbt/mlxtend/blob/7ed21f6c5df108102c26d8cdb3c4b899e007be61/mlxtend/classifier/stacking_cv_classification.py#L258
My concern would be it might be a bit confusing for users since it's mentioned in the doc as:
`Note: fit_base_estimators=False will enforce use_clones to be False,
        and is incompatible to most scikit-learn wrappers!
        For instance, if any form of cross-validation is performed
        this would require the re-fitting classifiers to training folds, which
        would raise a NotFitterError if fit_base_estimators=False.`

Hope my point is clear. Please take a look at the PR and I will continue with StackingRegressor. Possible another for StackingCV*, and you can decide if you want to include the change or not =)
